### PR TITLE
Add: Users API

### PR DESCRIPTION
**Pull request by firefinchdev**

### CODE REVIEW

**Comment by firefinchdev** [comment]
@championswimmer 
Pls Review.

File: src/controllers/user.js [hunk 1]
```diff
@@ -21,8 +21,16 @@ function findUserForTrustedClient(trustedClient, userId) {
   });
 }
 
+function findAllUsers(trustedClient, where) {
```
Comment (championswimmer) [inline src/controllers/user.js]: `findAllUsersWithFilter(trustedClient, filterArgs)`

File: src/routers/api/users.js [hunk 2]
```diff
@@ -184,4 +184,83 @@ router.get('/:id/address',
...
+          users = [users]
+      }
+      res.send(users)
+    } catch (error) {
+      res.send('Unknown user or unauthorized request')
+    }
+  }
+)
+
+function generateWhere(query) {
```
Comment (championswimmer) [inline src/routers/api/users.js]: move this function to controller. router layer should not have sequelize-specific code

File: src/controllers/user.js (hunk 1 above)
Comment (championswimmer) [reply src/controllers/user.js]: the filterArgs should come from req.query - only the required values

**Comment by firefinchdev** [comment]
@championswimmer Changes Done. Pls review

File: src/controllers/user.js [hunk 3]
```diff
@@ -21,8 +21,68 @@ function findUserForTrustedClient(trustedClient, userId) {
   });
 }
 
+function findAllUsersWithFilter(trustedClient, filterArgs) {
+  return User.findAll({
+    attributes: trustedClient ? undefined : ["id", "username", "email", "firstname", "lastname", "mobile_number"],
+    where: filterArgs || {},
```
Comment (championswimmer) [inline src/controllers/user.js]: `where: generateFilter(filterArgs)`

File: src/controllers/user.js [hunk 4]
```diff
@@ -21,8 +21,68 @@ function findUserForTrustedClient(trustedClient, userId) {
...
+  });
+}
+
+function generateFilter(query) {
```
Comment (championswimmer) [inline src/controllers/user.js]: `function generateFilter(filterArgs) {`

File: src/controllers/user.js [hunk 5]
```diff
@@ -21,8 +21,68 @@ function findUserForTrustedClient(trustedClient, userId) {
...
+  return whereObj
+}
+
 module.exports = {
   findUserById,
   updateUser,
-  findUserForTrustedClient
+  findUserForTrustedClient,
+  findAllUsersWithFilter,
+  generateFilter
```
Comment (championswimmer) [inline src/controllers/user.js]: dont export generateFilter

File: src/routers/api/users.js [hunk 6]
```diff
@@ -9,10 +9,40 @@ const passport = require('../../passport/passporthandler')
 const models = require('../../db/models').models
 
 const Raven = require('raven');
-const { findUserById , findUserForTrustedClient} = require('../../controllers/user');
+const sequelize = require('sequelize');
+const { findUserById , findUserForTrustedClient, findAllUsersWithFilter, generateFilter} = require('../../controllers/user');
```
Comment (championswimmer) [inline src/routers/api/users.js]: don't import generateFilter

File: src/routers/api/users.js [hunk 7]
```diff
@@ -9,10 +9,40 @@ const passport = require('../../passport/passporthandler')
...
+  passport.authenticate('bearer', {session: false}),
+  async function (req, res) {
+    // Send the user his own object if the token is user scoped
+    if (req.user && !req.authInfo.clientOnly && req.user.id) {
+      if (req.params.id == req.user.id) {
+        return res.send(req.user)
+      }
+    }
+
+    let whereObj = generateFilter(req.query)
```
Comment (championswimmer) [inline src/routers/api/users.js]: first validate that userid is numeric, not string etc and other things that you feel required.

then make this
```
filterArgs = {
  username: req.query.username,
  ...
}
then pass `filterArgs` into `findAllUsersWithFilter`
```

File: src/controllers/user.js [hunk 8]
```diff
@@ -21,8 +22,68 @@ function findUserForTrustedClient(trustedClient, userId) {
...
+    }
+  }
+  if (filterArgs.email) {
+    let email = filterArgs.email
+    // email = email.split('@')
+    // email[0] = email[0].split('').filter(c => !(c === '.')).join('')
+    // email = email.join('@')
+    // console.log('FF', email)
+    // whereObj.email = email
+    whereObj.email = sequelize.where(sequelize.fn('replace', sequelize.col('email'), '.', ''), sequelize.fn('replace', email, '.', ''))
```
Comment (championswimmer) [inline src/controllers/user.js]: add a comment as to what this does.

**Comment by championswimmer** [comment]
accepting this PR as the valid solution to the issue, please add a few comments where noted.

File: src/controllers/user.js [hunk 9]
```diff
@@ -21,8 +22,68 @@ function findUserForTrustedClient(trustedClient, userId) {
...
+  }
+  if (filterArgs.lastname) {
+    whereObj.lastname = {
+      $iLike: `${filterArgs.lastname}%`
+    }
+  }
+  if (filterArgs.email) {
+    let email = filterArgs.email
+    // email = email.split('@')
+    // email[0] = email[0].split('').filter(c => !(c === '.')).join('')
```
Comment (championswimmer) [inline src/controllers/user.js]: remove the commented old lines

**Pull request merged**